added productRepository like argument to constructor TelegramSender

Mansonita opened the pull request.

Commit:
added productRepository like argument to constructor TelegramSender

Commit:
i dont know whats the type for brand + category ( i should return when i  make request to mongoDB  in GetFilteredProducts methods

Commit:
added the class Result of Search which is responsible for checking  and obtainng result  from DB. And added new methods to it in UserStateRepository

kapitan123 commented on src/YourEasyRent/DataBase/UserStateRepository.cs:
```diff
@@ -55,8 +57,44 @@ public async Task<MenuStatus> GetCurrentStateForUser(string userId)
+            var requiredFields = new List<string> {"UserId", "Brand", "Category" };
+            foreach(var field in requiredFields)
+            {
+                var property = typeof(UserSearchStateDTO).GetProperty(field);
```
Вообще - я впечатлён. Хоть так и не делается, но то, что ты смогла правильно применить достаточно продвиную технику это круто.

kapitan123 replied:
Это работает правильно, но есть проблемы. Во-первых это медлено, во-вторых, в поиске места где используются .UserId - это место не попадёт.

Сделатьй просто 
if(dto.UserId == null)
return false;

kapitan123 commented on src/YourEasyRent/DataBase/UserStateRepository.cs:
```diff
@@ -55,8 +57,44 @@ public async Task<MenuStatus> GetCurrentStateForUser(string userId)
+
+            var listWithResult = new List<string>
+            {
+                brandAndCategoryResult.AsString
```
По идее тебе нужен отдельный тип здесь.
BrandAndCategory, либо можно вернуть Dictionary[string,string], ещё можно вернуть кортеж List<(string,stringP)>.

kapitan123 commented on src/YourEasyRent/Services/TelegramCallbackHandler.cs:
```diff
@@ -78,46 +78,69 @@ public async Task HandleUpdateAsync(TgButtonCallback tgButtonCallback)
+            }
+            if (tgButtonCallback.IsValueProductButton)//  протестировать как отрабатывает этот метод
+            {
+                var productButton = tgButtonCallback.GetProductButton();
```
У тебя, вроде был метод IsBrand

kapitan123 commented on src/YourEasyRent/Services/TelegramCallbackHandler.cs:
```diff
@@ -78,46 +78,69 @@ public async Task HandleUpdateAsync(TgButtonCallback tgButtonCallback)
-                //    MenuStatus status = MenuStatus.BrandMenu;
-                //    userSearchState.AddStatusToHistory(status);
-                //    //await _userStateRepository.UpdateAsync(userSearchState);
+                    var brand = tgButtonCallback.GetProductButton();
```
Чот я запутался. brand или product?

kapitan123 commented on src/YourEasyRent/Services/TelegramCallbackHandler.cs:
```diff
@@ -78,46 +78,69 @@ public async Task HandleUpdateAsync(TgButtonCallback tgButtonCallback)
 
-                //    await _telegramSender.SendBrandMenu(chatId); // тут еще нужно устанавливать и сохранять значение Статуса меню, чтобы потом можно было пользоваться кнопкой назад
-                //    return;
+                    userSearchState.ToMongoRepresintation();
```
userSearchState.ToMongoRepresintation(); по идее должен быть внутри репозитория самого

kapitan123 commented on src/YourEasyRent/Services/TgButtonCallback.cs:
```diff
@@ -12,7 +12,7 @@ public class TgButtonCallback
         public bool IsValidMessage => IsValidMsg();
         public bool IsValueMenuMessage  => IsMenuButton();
-        public bool IsBrandMenu { get; set; }
+        public bool IsBrandMenu { get; set; }//  не знаю как мне сделать проверку для разных меню, чтобы потом воспользоваться этими свойствами в Handlder
```
А чего тама разного?

kapitan123 commented on src/YourEasyRent/UserState/ResultOfSearch.cs:
```diff
@@ -0,0 +1,51 @@
+        private readonly IProductRepository _productRepository;
+        public bool IsReadyToSearch { get; private set; }
+
+        public ResultOfSearch(IUserStateRepository userStateRepository, IProductRepository productRepository)
```
Неплохо вполне!
Но мне кажется все эти методы можно расскидать по уже существующим классам
IsReadyToSearchMethod в UserSearchState

GetFilteredProductsMessage мне кажется можно перенести в tgSender

kapitan123 committed:
Tst

Commit:
added method GetFilteredProductsForSearch into USRepository which return the Tuple with brand and category results.
added methods SendResults and GetFilteredProductsMessage into class TelegramSender. Brushed up other part of the code.

Commit:
added new files

Commit:
deleted errors

Commit:
remaned ckass Product

Commit:
removed extra lines
